mlflow/utils/virtualenv.py

```diff
@@ -0,0 +1,273 @@
+    environment.
+
+    :param local_model_path: Local directory containing the model artifacts.
+    :return: `_PythonEnv` instance.
+    """
+    python_env_file = local_model_path / _PYTHON_ENV_FILE_NAME
+    requirements_file = local_model_path / _REQUIREMENTS_FILE_NAME
+    conda_env_file = local_model_path / _CONDA_ENV_FILE_NAME
+    if python_env_file.exists():
+        return _PythonEnv.from_yaml(python_env_file)
+    else:
+        _logger.info(
+            "This model is missing %s, which is because it was logged in an older version"
+            "of MLflow (< 1.26.0) that does not support restoring a model environment with "
+            "virtualenv. Attempting to extract model dependencies from %s and %s instead.",
+            _PYTHON_ENV_FILE_NAME,
+            _REQUIREMENTS_FILE_NAME,
+            _CONDA_ENV_FILE_NAME,
+        )
+        conda_deps = _get_conda_dependencies(conda_env_file)
+        build_packages = ("python", *_PythonEnv.BUILD_PACKAGES)
+        conda_deps = [d for d in conda_deps if _get_package_name(d) not in build_packages]
+        if conda_deps:
+            _logger.warning(
+                f"This model contains conda dependencies: {conda_deps}. The resulting virtualenv "
+                "environment will not install them and may not be able to load the model."
+            )
+        if requirements_file.exists():
+            deps = _PythonEnv.get_dependencies_from_conda_yaml(conda_env_file)
+            return _PythonEnv(
+                python=deps["python"],
+                build_dependencies=deps["build_dependencies"],
+                dependencies=[f"-r {_REQUIREMENTS_FILE_NAME}"],
+            )
+        else:
+            return _PythonEnv.from_conda_yaml(conda_env_file)
+
+
+def _create_virtualenv(local_model_path, python_bin_path, env_dir, python_env):
+    # Created a command to activate the environment
+    paths = ("bin", "activate") if _IS_UNIX else ("Scripts", "activate.bat")
+    activate_cmd = env_dir.joinpath(*paths)
+    activate_cmd = f"source {activate_cmd}" if _IS_UNIX else activate_cmd
+
+    if env_dir.exists():
+        _logger.info("Environment %s already exists", env_dir)
+        return activate_cmd
+
+    _logger.info("Creating a new environment %s", env_dir)
+    _exec_cmd(["virtualenv", "--python", python_bin_path, env_dir], capture_output=False)
+
+    _logger.info("Installing dependencies")
+    for deps in filter(None, [python_env.build_dependencies, python_env.dependencies]):
+        # Use a unique name to avoid conflicting custom requirements files logged by a user
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Create a temporary requirements file
+            tmpdir = Path(tmpdir)
+            tmp_req_file = tmpdir.joinpath(f"requirements.{uuid.uuid4().hex}.txt")
+            tmp_req_file.write_text("\n".join(deps))
+            # Copy `requirements.txt` and `constraints.txt` in the temporary directory.
```

Comment:
There could still be other files beyond `requirements.txt` and `constraints.txt` right? E.g. a requirements file that references another requirements file. Can we handle this?

Reply:
Offline discussion: use the initial approach (= perform package installation in the model directory)